### PR TITLE
[docs] update io manager page to address deps api changes

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
updates the io manager section of the docs to discuss deps a bit and make I/O managers seem more like a choice than a requirement

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15422** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15422" target="_blank">Graphite</a> 
    * **PR #15438** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15438" target="_blank">Graphite</a> 
      * **PR #15453** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15453" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15453?utm_source=stack-comment).

File: docs/content/concepts/io-management/io-managers.mdx [hunk 1]
```diff
@@ -82,6 +82,33 @@ Dagster provides out-of-the-box I/O managers for popular storage systems: AWS S3
...
+
+Other reasons you may not want to use I/O managers:
+
+- You want to run a SQL query that creates or updates a table in a database (the above example).
+- Your pipeline manages I/O on its own by using other libraries/tools that write to storage.
+- You have unique and custom storage requirements for each asset.
+- Your assets won't fit in memory (for example, a database table with billions of rows).
+- You have an existing pipeline that manages its own I/O and want to run it with Dagster with minimal code changes.
+- You simply prefer to have the reading and writing code explicitly defined in each asset.
+
```
Comment (jamiedemaria) [inline docs/content/concepts/io-management/io-managers.mdx:109]: Do you think its worth having a companion list of scenarios when I/O managers would be helpful?

File: docs/content/concepts/io-management/io-managers.mdx [hunk 2]
```diff
@@ -82,6 +82,33 @@ Dagster provides out-of-the-box I/O managers for popular storage systems: AWS S3
 
 I/O managers are provided through the [resources](/concepts/resources) system which means you can supply different I/O managers for the same assets or ops in different situations. For example, you might use an in-memory I/O manager for unit-testing and an Amazon S3 I/O manager in production.
 
+
+### When to use I/O managers
+
+I/O managers are a powerful tool that can simplify your code and allow you to easily modify where your assets are stored. However, they are not required, nor are they the best option, in all scenarios.
+
+Some assets do not need an I/O manager. For example, you may write an asset that executes a Snowflake query that creates a new table based on the data of another table. This asset would depend on the existing table, but does not need to load that table in memory in order to execute the query:
```
Comment (erinkcochran87) [inline docs/content/concepts/io-management/io-managers.mdx:90]:
```suggestion
Some assets do not need an I/O manager. For example, an asset that executes a Snowflake query to create a new table based on the data of another table. This asset would depend on the existing table, but does not need to load that table in memory in order to execute the query:
```

Suggesting some tweaks to the second sentence to make it more concise

File: docs/content/concepts/io-management/io-managers.mdx (hunk 1 above)
Comment (erinkcochran87) [reply docs/content/concepts/io-management/io-managers.mdx:109]: Perhaps - what would that list look like?

File: docs/content/concepts/io-management/io-managers.mdx [hunk 3]
```diff
@@ -82,6 +82,33 @@ Dagster provides out-of-the-box I/O managers for popular storage systems: AWS S3
...
+Other reasons you may not want to use I/O managers:
+
+- You want to run a SQL query that creates or updates a table in a database (the above example).
+- Your pipeline manages I/O on its own by using other libraries/tools that write to storage.
+- You have unique and custom storage requirements for each asset.
+- Your assets won't fit in memory (for example, a database table with billions of rows).
+- You have an existing pipeline that manages its own I/O and want to run it with Dagster with minimal code changes.
+- You simply prefer to have the reading and writing code explicitly defined in each asset.
+
+As a general rule, if you find yourself getting tripped up writing your pipeline to use I/O managers, or you find yourself making your pipeline more complicated in order to use I/O managers, you will likely have more success not using I/O managers. In these cases you should use [`deps`](/concepts/assets/software-defined-assets#defining-basic-dependencies) to define dependencies.
```
Comment (erinkcochran87) [inline docs/content/concepts/io-management/io-managers.mdx:110]:
```suggestion
As a general rule, if you get confused when implementing I/O managers or your pipeline becomes more complicated in order to use them, we recommend **not** using them. In these cases, use [`deps`](/concepts/assets/software-defined-assets#defining-basic-dependencies) to define dependencies.
```

Thoughts on making this less wordy?

File: docs/content/concepts/io-management/io-managers.mdx (hunk 1 above)
Comment (jamiedemaria) [reply docs/content/concepts/io-management/io-managers.mdx:109]: off the top of my head 
- there is consistency with how you want to store multiple assets (in the same folder, in the same db schema following the same naming conventions)
- you want to be able to swap IO managers for testing, or other environments 
- you always want to load the upstream asset into memory in downstream assets 

maybe a good test we could mention to users is that if they are writing the same code at the top and bottom of each asset function to load the upstream dependencies and save the asset, then factoring that out to an I/O manager could be a good idea

Comment (jamiedemaria) [reply docs/content/concepts/io-management/io-managers.mdx:112]: i like the idea of moving it up

**Review by erinkcochran87 — APPROVED** [review]

**Pull request merged**